crates/cast/src/lib.rs

```diff
@@ -1448,23 +1448,66 @@ impl SimpleCast {
         Ok(ParseUnits::parse_units(value, unit)?.to_string())
     }
 
-    /// Decodes rlp encoded list with hex data
-    ///
-    /// # Example
+    // Decodes RLP encoded data with validation for canonical integer representation
     ///
+    /// # Examples
     /// ```
     /// use cast::SimpleCast as Cast;
     ///
-    /// assert_eq!(Cast::from_rlp("0xc0").unwrap(), "[]");
-    /// assert_eq!(Cast::from_rlp("0x0f").unwrap(), "\"0x0f\"");
-    /// assert_eq!(Cast::from_rlp("0x33").unwrap(), "\"0x33\"");
-    /// assert_eq!(Cast::from_rlp("0xc161").unwrap(), "[\"0x61\"]");
-    /// assert_eq!(Cast::from_rlp("0xc26162").unwrap(), "[\"0x61\",\"0x62\"]");
+    /// assert_eq!(Cast::from_rlp("0xc0", false).unwrap(), "[]");
+    /// assert_eq!(Cast::from_rlp("0x0f", false).unwrap(), "\"0x0f\"");
+    /// assert_eq!(Cast::from_rlp("0x33", false).unwrap(), "\"0x33\"");
+    /// assert_eq!(Cast::from_rlp("0xc161", false).unwrap(), "[\"0x61\"]");
+    /// assert_eq!(Cast::from_rlp("820002", true).is_err(), true);
+    /// assert_eq!(Cast::from_rlp("820002", false).unwrap(), "\"0x0002\"");
+    /// assert_eq!(Cast::from_rlp("00", true).is_err(), true);
+    /// assert_eq!(Cast::from_rlp("00", false).unwrap(), "\"0x00\"");
     /// # Ok::<_, eyre::Report>(())
     /// ```
-    pub fn from_rlp(value: impl AsRef<str>) -> Result<String> {
+    pub fn from_rlp(value: impl AsRef<str>, as_int: bool) -> Result<String> {
+        fn validate_canonical_int(bytes: &[u8]) -> Result<()> {
+            if bytes.is_empty() {
+                return Ok(());
+            }
+
+            match bytes.len() {
+                1 => {
+                    if bytes[0] < 0x80 {
+                        eyre::bail!("rlp: non-canonical size information")
+                    }
+                }
+                _ => {
+                    if bytes[0] == 0 {
+                        eyre::bail!("rlp: non-canonical integer (leading zero bytes)")
+                    }
+                    if bytes[0] >= 0x80 {
+                        let size: usize = (bytes[0] - 0x80).into();
+
+                        // Size tags must use the smallest possible encoding.
+                        if size > bytes.len() - 1 {
+                            eyre::bail!("rlp: non-canonical size information")
+                        }
+
+                        if bytes[1] == 0 {
+                            eyre::bail!("rlp: non-canonical size information (leading zero bytes)")
+                        }                        
+                    }
+                }
+            }
+
+            Ok(())
+        }
+
         let bytes = hex::decode(value.as_ref()).wrap_err("Could not decode hex")?;
+
+        if as_int {
+            validate_canonical_int(&bytes).wrap_err("Non-canonical integer")?;
+            let uint_value = U256::from_str(value.as_ref())?;
+            return Ok(format!("{uint_value:#?}"));
+        }
+
```

Comment:
sorry I mean decode rlp as uint (just `U256::decode`), that way we don't need the extra validation because the decoding would just fail if it's an incorrect integer

Reply:
thank you so much, I've updated